platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,233 @@
+            // active roster present in the state or throw if not found.
+            return Objects.requireNonNull(
+                    lastUsedActiveRoster, "Active Roster must be present in the state during normal network restart.");
+        }
+
+        // software upgrade is detected, adopt the candidate roster as the active roster.
+        final Roster activeRoster = adoptCandidateRoster(initialState.get().getRound() + 1);
+        commit();
+        return activeRoster;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * Adopts the candidate roster present in the state as the active roster.
+     *
+     * @param roundNumber the round number in which the candidate roster became active
+     */
+    @NonNull
+    private Roster adoptCandidateRoster(final long roundNumber) {
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Roster candidateRoster = rosterMap.get(ProtoBytes.newBuilder()
+                .value(previousRosterState.candidateRosterHash())
+                .build());
+        if (candidateRoster == null) {
+            throw new IllegalStateException("Candidate roster not found in the state.");
+        }
+
+        storeAsActive(candidateRoster, roundNumber);
+        return candidateRoster;
+    }
+
+    /**
+     * Stores this roster as the active roster.
+     *
+     * @param roster        a roster to set as active
+     * @param round        the round in which this roster became active
+     */
+    private void storeAsActive(@NonNull final Roster roster, final long round) {
+        Objects.requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        final RosterState previousRosterState = rosterStateOrThrow();
+
+        // update the roster state
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        final Bytes activeRosterHash = RosterUtils.hash(roster).getBytes();
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        // remove the formerly previous active roster, i.e., the roster that was active before the last two adopted
+        // rosters, if any.
+        if (roundRosterPairs.size() > 2) {
+            final RoundRosterPair lastRemovedStillInState = roundRosterPairs.removeLast();
+
+            // remove the roster from the map if it is still present in the Roster state for whatever reason.
+            final boolean rosterStateDuplicateFound = roundRosterPairs.stream()
+                    .anyMatch(peer -> peer.activeRosterHash().equals(lastRemovedStillInState.activeRosterHash()));
```

Comment:
Did you mean to name the lambda variable `pair` instead?

Reply:
I don't think it's possible either but there was a feedback from @litt3 that since it's a list which can have duplicates, it's worth doing the additional check and removal too (hence my 'for whatever reason' in the comment).

Reply:
If the roster being adopted, `x`, is a duplicate of roster `x-2`, then blindly removing roster `x-2` from the map would entail also deleting roster `x`, which is still needed. The is both technically possible and reasonable, in cases e.g. where a node joins the network, and then leaves, without any other membership changes.

Reply:
Good point - this should be explained in the java doc

Reply:
Appears to be resolved, marking resolved.